### PR TITLE
chore: use blsttc

### PR DESCRIPTION
**Pull request by RolandSherwin**
Uses our own implementation of TC

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -17,3 +17,4 @@ rand = "0.6.5"
 serde = { version = "1.0.102", features = ["derive", "rc"] }
 threshold_crypto = { rev = "624eeee", git = "https://github.com/poanetwork/threshold_crypto" }
```
Comment (grumbach) [inline Cargo.toml]: Do we still need this?

File: src/sdkg.rs [hunk 2]
```diff
@@ -178,11 +134,11 @@ impl From<bincode::Error> for Error {
     }
 }
 
-impl Error {
-    fn encrypt<E: ToString>(err: E) -> Error {
-        Error::Encrypt(err.to_string())
-    }
-}
+// impl Error {
```
Comment (grumbach) [inline src/sdkg.rs:137]: Why did you comment this out?

File: src/sdkg.rs [hunk 3]
```diff
@@ -343,7 +299,7 @@ impl<N: NodeIdT, PK: PublicKey> SyncKeyGen<N, PK> {
     ///
     /// All participating nodes must handle the exact same sequence of messages.
     /// Note that `handle_part` also needs to explicitly be called with this instance's own `Part`.
-    pub fn handle_part<R: rand::Rng>(
+    pub fn handle_part<R: bls::rand::RngCore>(
```
Comment (grumbach) [inline src/sdkg.rs:302]: Why use `bls::rand::RngCore` instead of `rand::Rng`?

File: src/sdkg.rs [hunk 4]
```diff
@@ -498,11 +455,11 @@ impl<N: NodeIdT, PK: PublicKey> SyncKeyGen<N, PK> {
         let ser_val = self
             .sec_key
             .decrypt(&values[our_idx as usize])
-            .map_err(|_| AckFault::DecryptValue)?;
+            .ok_or(AckFault::DecryptValue)?;
         let val = bincode::deserialize::<FieldWrap<Fr>>(&ser_val)
             .map_err(|_| AckFault::DeserializeValue)?
             .into_inner();
-        if part.commit.evaluate(our_idx + 1, sender_idx + 1) != G1Affine::one().mul(val) {
```
Comment (grumbach) [inline src/sdkg.rs:505]: I'm not knowledgable about this but is `one()` here the same as `generator()`?

**Review by grumbach — COMMENTED** [review]
Great job! Couple of comments and questions but looking good!

File: src/sdkg.rs (hunk 4 above)
Comment (RolandSherwin) [reply src/sdkg.rs:505]: I'm not knowledgeable as well :sweat_smile: , but just followed the function call and `one()` pointed me to a macro called `cruve_impl!` 
image
and I found that `one()` returns `CurveAffine::get_generator()`. And our generator is from the trait `blsttc::group::prime::PrimeCurveAffine`, so it must be the one right?
image

File: src/sdkg.rs (hunk 2 above)
Comment (RolandSherwin) [reply src/sdkg.rs:137]: The old threshold crypto lib's `encrypt` with rng returned a `Result` enum while our `encrypt_with_rng()` is a non-result type. Thus the variant ` Error::Encrypt` was unused.

File: src/sdkg.rs [hunk 5]
```diff
@@ -70,90 +70,46 @@
...
-        msg: M,
-        rng: &mut R,
-    ) -> Result<Vec<u8>, bincode::Error> {
-        bincode::serialize(&self.encrypt_with_rng(rng, msg))
-    }
-}
-
 /// A map assigning to each node ID a public key, wrapped in an `Arc`.
-pub type PubKeyMap<N, PK = threshold_crypto::PublicKey> = Arc<BTreeMap<N, PK>>;
+pub type PubKeyMap<N, PK = PublicKey> = Arc<BTreeMap<N, PK>>;
```
[the remark this answers is not in the document]
Comment (RolandSherwin) [reply src/sdkg.rs:96]: Got it! Will make changes!

[the remark this answers is not in the document]
Comment (RolandSherwin) [reply src/sdkg.rs:96]: As discussed in the huddle, I will leave the `NodeIdT` trait as it as for now, until we decide on which Id to use.

**Review by grumbach — APPROVED** [review]
LGTM

**Pull request merged**